src/config/transport/transport-strategy.ts

```diff
@@ -0,0 +1,85 @@
+import {IGrpcConfiguration} from './grpc-configuration';
+
+export interface ITransportStrategy {
+  getMaxConcurrentRequests(): number | null;
+
+  getGrpcConfig(): IGrpcConfiguration;
+
+  // TODO: for use in middleware
+  withMaxConcurrentRequests(maxConcurrentRequests: number): ITransportStrategy;
+
+  withGrpcConfig(grpcConfig: IGrpcConfiguration): ITransportStrategy;
+
+  withClientTimeout(clientTimeout: number): ITransportStrategy;
+}
+
+export class StaticGrpcConfiguration implements IGrpcConfiguration {
+  private readonly deadlineMilliseconds: number;
+  private readonly maxSessionMemory: number;
+  constructor(deadlineMilliseconds: number, maxSessionMemory: number) {
+    this.deadlineMilliseconds = deadlineMilliseconds;
+    this.maxSessionMemory = maxSessionMemory;
+  }
+
+  getDeadlineMilliseconds(): number {
+    return this.deadlineMilliseconds;
+  }
+
+  getMaxSessionMemory(): number {
+    return this.maxSessionMemory;
+  }
+
+  withDeadlineMilliseconds(
+    deadlineMilliseconds: number
+  ): StaticGrpcConfiguration {
+    return new StaticGrpcConfiguration(
+      deadlineMilliseconds,
+      this.maxSessionMemory
+    );
+  }
+
+  withMaxSessionMemory(maxSessionMemory: number): StaticGrpcConfiguration {
+    return new StaticGrpcConfiguration(
+      this.deadlineMilliseconds,
+      maxSessionMemory
+    );
+  }
+}
+
+export class StaticTransportStrategy implements ITransportStrategy {
+  private readonly maxConcurrentRequests: number | null;
+  private readonly grpcConfig: IGrpcConfiguration;
+
+  constructor(
+    maxConcurrentRequests: number | null,
+    grpcConfiguration: IGrpcConfiguration
+  ) {
+    this.maxConcurrentRequests = maxConcurrentRequests;
+    this.grpcConfig = grpcConfiguration;
+  }
+
+  getMaxConcurrentRequests(): number | null {
+    return this.maxConcurrentRequests;
+  }
+
+  getGrpcConfig(): IGrpcConfiguration {
+    return this.grpcConfig;
+  }
+
+  withMaxConcurrentRequests(
+    maxConcurrentRequests: number
+  ): StaticTransportStrategy {
+    return new StaticTransportStrategy(maxConcurrentRequests, this.grpcConfig);
+  }
+
+  withGrpcConfig(grpcConfig: IGrpcConfiguration): StaticTransportStrategy {
+    return new StaticTransportStrategy(this.maxConcurrentRequests, grpcConfig);
+  }
+
+  withClientTimeout(clientTimeout: number): StaticTransportStrategy {
```

Comment:
It's a little confusing that "clientTimeout" sets "deadlineMilliseconds".

Should deadlineMilliseconds be timeoutMilliseconds? Should clientTimeout be clientTimeoutMilliseconds?

Reply:
this is the layer at which we are trying to abstract away grpc implementation details.

in grpc-speak the correct name for this field is `*Deadline*`.

In customer-facing APIs we usually don't want them to have to think about gRPC unless they really dig down into the guts, so we expose it as `ClientTimeout`.  This would theoretically let us use this config item for other transports as well.

I think you are correct that `clientTimeout` should be `clientTimeoutMilliseconds` though, unless there is a sane duration object we can use in JS-land.  I will change this in one of those two ways.

Reply:
That's fair if the idea is to keep GRPC terms at the GRPC layer.

There doesn't seem to be a built-in Javascript nor Typescript duration object. I found [this popular module](https://www.npmjs.com/package/duration), but it doesn't let you construct Durations out of minutes and seconds. 🤷

We start building up a style guide with things like how to handle time durations in various languages and how to name args that take them.